File: lib/models/message.dart
class Message {
  final int id;
  final int senderId;
  final int receiverId;
  final String content;
  final String type;
  final int timestamp;
  final bool isRead;
  final int? hikeId;
  final int? senderHikeId;
  final int? receiverHikeId;
  /// 好友消息中的图片/附件 URL（如 type 为 image 时）
  final String? attachmentUrl;

  Message({
    required this.id,
    required this.senderId,
    required this.receiverId,
    required this.content,
    required this.type,
    required this.timestamp,
    required this.isRead,
    this.hikeId,
    this.senderHikeId,
    this.receiverHikeId,
    this.attachmentUrl,
  });

  factory Message.fromJson(Map<String, dynamic> json) {
    return Message(
      id: json['id'] ?? json['remote_id'] ?? json['local_id'] ?? 0,
      senderId: json['sender_id'],
      receiverId: json['receiver_id'] ?? 0, // handle optional
      content: json['content'],
      type: json['type'],
      timestamp: json['timestamp'],
      isRead: (json['is_read'] is int) ? (json['is_read'] == 1) : (json['is_read'] ?? false),
      hikeId: json['hike_id'],
      senderHikeId: json['sender_hike_id'],
      receiverHikeId: json['receiver_hike_id'],
      attachmentUrl: json['attachment_url'],
    );
  }

  Map<String, dynamic> toJson() {
    final m = <String, dynamic>{
      'id': id,
      'sender_id': senderId,
      'receiver_id': receiverId,
      'content': content,
      'type': type,
      'timestamp': timestamp,
      'is_read': isRead ? 1 : 0,
      'hike_id': hikeId,
      'sender_hike_id': senderHikeId,
      'receiver_hike_id': receiverHikeId,
    };
    if (attachmentUrl != null) m['attachment_url'] = attachmentUrl;
    return m;
  }
}

File: backend/app/models/message.py
from sqlalchemy import Column, Integer, String, Float, ForeignKey, Boolean, BigInteger, UniqueConstraint
from sqlalchemy.orm import relationship
from app.database.database import Base

class Message(Base):
    __tablename__ = "messages"

    id = Column(Integer, primary_key=True, index=True)
    sender_id = Column(Integer, ForeignKey("users.id"))
    receiver_id = Column(Integer, ForeignKey("users.id"), nullable=True) # Null for broadcast/group
    
    content = Column(String)
    type = Column(String) # 'text', 'image', 'sos', 'feedback', 'question'
    timestamp = Column(BigInteger)
    is_read = Column(Boolean, default=False)
    
    sender_hike_id = Column(Integer, ForeignKey("hiking_records.id"), nullable=True)
    receiver_hike_id = Column(Integer, ForeignKey("hiking_records.id"), nullable=True)
    
    sender = relationship("User", foreign_keys=[sender_id])
    receiver = relationship("User", foreign_keys=[receiver_id])


class FriendMessage(Base):
    """好友间消息：仅保存两人成为好友后的对话（文字、图片、emoji 等）。"""
    __tablename__ = "friend_messages"

    id = Column(Integer, primary_key=True, index=True)
    sender_id = Column(Integer, ForeignKey("users.id"))
    receiver_id = Column(Integer, ForeignKey("users.id"))

    content = Column(String)  # 文本内容或 emoji
    type = Column(String, default="text")  # text, image, emoji
    attachment_url = Column(String, nullable=True)  # 图片等附件的 URL
    timestamp = Column(BigInteger)
    is_read = Column(Boolean, default=False)

    sender = relationship("User", foreign_keys=[sender_id])
    receiver = relationship("User", foreign_keys=[receiver_id])


class TempFriendship(Base):
    """临时会话关系：仅用于非好友期间的临时沟通会话列表。"""
    __tablename__ = "temp_friendships"

    id = Column(Integer, primary_key=True, index=True)
    user_id = Column(Integer, ForeignKey("users.id"), index=True)
    partner_id = Column(Integer, ForeignKey("users.id"), index=True)
    last_message = Column(String, nullable=True)
    last_message_type = Column(String, default="text")
    last_timestamp = Column(BigInteger)
    created_at = Column(BigInteger)
    updated_at = Column(BigInteger)

    __table_args__ = (
        UniqueConstraint("user_id", "partner_id", name="uq_temp_friendship_user_partner"),
    )

    user = relationship("User", foreign_keys=[user_id])
    partner = relationship("User", foreign_keys=[partner_id])


class Feedback(Base):
    __tablename__ = "feedbacks"
    
    id = Column(Integer, primary_key=True, index=True)
    user_id = Column(Integer, ForeignKey("users.id"))
    
    type = Column(String) # 'blocked', 'weather', etc.
    content = Column(String)
    latitude = Column(Float)
    longitude = Column(Float)
    address = Column(String)
    photos = Column(String, nullable=True) # JSON list of URLs
    
    created_at = Column(BigInteger)
    status = Column(String, default='ACTIVE') # ACTIVE, EXPIRED
    
    view_count = Column(Integer, default=0)
    confirm_count = Column(Integer, default=0)
    forward_count = Column(Integer, default=0)
    
    user = relationship("User", back_populates="feedbacks")

class SOSAlert(Base):
    __tablename__ = "sos_alerts"
    
    id = Column(Integer, primary_key=True, index=True)
    user_id = Column(Integer, ForeignKey("users.id"))
    latitude = Column(Float)
    longitude = Column(Float)
    message = Column(String)
    photos = Column(String, nullable=True)  # JSON list of base64 strings (临时存DB，后续可换成URL列表)
    status = Column(String, default='ACTIVE') # ACTIVE, RESOLVED
    created_at = Column(BigInteger)
    resolved_at = Column(BigInteger, nullable=True)
    
    user = relationship("User")


class FeedbackConfirm(Base):
    __tablename__ = "feedback_confirms"

    id = Column(Integer, primary_key=True, index=True)
    feedback_id = Column(Integer, ForeignKey("feedbacks.id"), index=True)
    user_id = Column(Integer, ForeignKey("users.id"), index=True)
    created_at = Column(BigInteger)

    __table_args__ = (
        UniqueConstraint("feedback_id", "user_id", name="uq_feedback_user_confirm"),
    )

    feedback = relationship("Feedback")
    user = relationship("User")


class FeedbackComment(Base):
    __tablename__ = "feedback_comments"

    id = Column(Integer, primary_key=True, index=True)
    feedback_id = Column(Integer, ForeignKey("feedbacks.id"), index=True)
    user_id = Column(Integer, ForeignKey("users.id"), index=True)
    content = Column(String)
    created_at = Column(BigInteger)

    feedback = relationship("Feedback")
    user = relationship("User")
